onnxruntime/core/graph/graph.cc

```diff
@@ -1498,7 +1515,9 @@ Status Graph::InferAndVerifySubgraphTypes(const Node& node, Graph& subgraph,
 
   // now that we have handled the input types, do the type/shape inferencing for the subgraph
   // to flow the type/shape info through it
-  status = subgraph.PerformTypeAndShapeInferencing();
+  // TODO: Handle override-type option correctly for subgraphs.
+  Graph::ResolveOptions options;
```

Comment:
Does this need to be fixed?

Reply:
@gramalingam can you comment on this?

---
In reply to: [397040118](https://github.com/microsoft/onnxruntime/pull/3174#discussion_r397040118) [](ancestors = 397040118)

Reply:
AFAIK, the training extensions have not been tested with any control-flow constructs. The override-type option is used to perform a graph-transformation that has the effect of changing the types of variables (for example, from float32 tensor to float16 tensor for lower-precision computation). As a result, after such a transformation, we can't perform a normal type-checking ... instead, if the inference figures out a new type for a variable (different from its existing type), it updates the variable's type instead of producing an error message. This extension has not been tested with control-flow constructs. Testing it with control-flow constructs may reveal if something more is required. I tried thinking through this, and can't add more specific information whether some extension is definitely required.

Reply:
Can a basic test be added using one of the control flow ops where the subgraph uses a value that has its type changed via this mechanism? I would expect outer scope values are more likely to just work, but explicit inputs to the subgraph may be trickier if the type of those change. 

---
In reply to: [401093531](https://github.com/microsoft/onnxruntime/pull/3174#discussion_r401093531) [](ancestors = 401093531)